src/AutoRest.CSharp/Common/Generation/Types/TypeFactory.cs

```diff
@@ -29,9 +29,12 @@ public TypeFactory(OutputLibrary library)
             _library = library;
         }
 
-        public CSharpType CreateType(InputType inputType) => inputType switch
+        public CSharpType CreateType(InputType inputType, bool readOnly = false) => inputType switch
         {
-            InputListType listType             => new CSharpType(typeof(IList<>), listType.IsNullable, CreateType(listType.ElementType)),
+            InputListType listType when readOnly is true                => new CSharpType(typeof(IReadOnlyList<>), listType.IsNullable, CreateType(listType.ElementType)),
```

Comment:
I'm not sure I like the idea of adding a `readOnly` parameter here because it is applicable only in a small subset of cases and requires a context that isn't available, e.g.:
- Read-only version of `IList<T>` is `IReadOnlyList<T>` for properties but `IEnumerable<T>` for method parameters
- Read-only version of  IList<IList<T>> is not well defined

Can you use `TypeFactory.GetOutputType` for your case?

Reply:
Agree. I also felt smell here, but I didn't know a better place to put the logic. Thanks for the information. I implemented a similar method in `ModelTypeProvider`, like `SchemaObjectType`: https://github.com/Azure/autorest.csharp/blob/f7ea672aeaa84049dfccc5c5e5ab6b93d3e7c86c/src/AutoRest.CSharp/Common/Output/Models/Types/SchemaObjectType.cs#L472-L483